fitz/utils.py

```diff
@@ -1834,6 +1834,141 @@ def insert_text(
+        orig_height = rect.y1  # original relevant height
+        mycss = "body {margin:1px;}" + css  # append user CSS
+        if isinstance(text, str):  # if a string, convert to a Story
+            story = Story(html=text, user_css=mycss, archive=archive)
+        else:
+            story = text  # assume we have a story
+
+        # ---------------------------------------------------------------------
+        # Loop for fitting text into rectangle.
+        # We will expand the rectangle until we have a fit in any case. Only
+        # finally we will check "adjust" and report the overflow amount if
+        # no adjustment was intended.
+        # ---------------------------------------------------------------------
+        more = True
+        while more:
+            story.reset()  # rewind story to the beginning
+            fp = io.BytesIO()  # use for file output
+            writer = DocumentWriter(fp)
+            dev = writer.begin_page(rect)  # make a temporary page
+            more, filled = story.place(rect)  # position text in the rect
+            if more:  # text did not fit in this clip, so enlarge and try again
+                rect *= 1.01  # enlarge by 1%
+                links = []
+                continue
+            # we stayed inside rect
+            # invoke position collector for identifying links
+            story.element_positions(collect, args={})
+            story.draw(dev)
+            writer.end_page()  # end the page
+            writer.close()  # close the writer
+            break  # leave the loop
+
+        try:
+            doc = Document("pdf", fp)  # try make a PDF from memory
+        except Exception as e:
+            # if Document creation failed, return unsuccessful insert
+            print(e)
+            return None, FZ_MAX_INF_RECT, Identity, []
+        if rect.y1 > orig_height and adjust is False:
+            # no fit and we were told to return a no-op
+            doc.close()
+            return None, rect.y1, Identity, []
+
+        # coming here means we have fitted the content in the rectangle
+        # we are returning the temp PDF and associated information
+        return (
+            doc,
+            filled[3] - filled[1],
+            (rect * Matrix(-rotate)).torect(clip),
+            links,
+        )
+
+    # normalize rotation angle
+    while rotate < 0:
+        rotate += 360
+    while rotate >= 360:
+        rotate -= 360
+    if not rotate % 90 == 0:
+        raise ValueError("bad rotation angle")
+    rect = Rect(rect)
+    if css is None:
+        css = ""
+
+    # layout text in rectangle of a temp PDF
+    doc, height, matrix, links = story_maker(
+        rect, text, adjust=adjust, rotate=rotate, css=css, archive=archive
+    )
+
+    rc = rect.height - height if rotate in (0, 180) else rect.width - height
+    if doc is not None:
+        if rc < 0:  # layout was successful: negative values can be ignored
+            rc = 0
```

Comment:
I'm a bit confused by this, and not sure this matches the docs which say that we return negative if unsuccessful, positive if successful.

`rc < 0` means that `rect.height - height < 0`, so `rect.height < height`, so `rect` is not big enough, contrary to the comment `layour was succesful`?

So we seem to be setting `rc` to zero if the layout failed, else positive if layout succeeded?